File: cosmoauthorization/src/java/it/csi/cosmo/cosmoauthorization/business/service/CertificatiService.java
/*
 * Copyright CSI-Piemonte - 2023
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

package it.csi.cosmo.cosmoauthorization.business.service;

import java.util.List;
import it.csi.cosmo.cosmoauthorization.dto.rest.CertificatoFirma;

/**
 *
 */

public interface CertificatiService {


  /**
   * Metodo che restituisce il certificato di cui si passa l'id
   *
   * @param idCertificato l'id del certificato che deve essere ricercato
   * @return il certificato identificato
   */
  CertificatoFirma getCertificato(String idCertificato);

  /**
   * Metodo che restituisce tutti i certificati di firma dell'utente in sessione
   *
   * @return una lista con tutti i certificati di firma dell'utente passato in argomento
   */
  List<CertificatoFirma> getCertificati();

  /**
   * Metodo per il primo salvataggio di un certificato di firma dell'utente in sessione
   *
   * @param body contiene tutte le informazioni necessarie di un certificato di firma
   * @return il certificato di firma salvato
   */
  CertificatoFirma postCertificato(CertificatoFirma body);


  /**
   * Metodo per l'aggiornamento di un certificato di firma dell'utente in sessione
   *
   * @param idCertificato del certificato da aggiornare
   * @param body contiene tutte le informazioni del certificato di firma da aggiornare
   * @return il certificato di firma aggiornato
   */
  CertificatoFirma putCertificato(String idCertificato, CertificatoFirma body);


  /**
   * Metodo per eliminare logicamente (quindi valorizzando data cancellazione e utente
   * cancellazione) del certificato di firma di cui si passa l'id
   *
   * @param idCertificato e' l'id del certificato che deve essere eliminato
   * @return il certificato di firma eliminato
   */
  CertificatoFirma deleteCertificato(String idCertificato);

  /**
   * Metodo che imposta l'ultimo certificato di firma utilizzato
   *
   * @param idCertificato e' l'id dell'ultimo certificato di firma utilizzato dall'utente
   * @return l'ultimo certificato di firma utilizzato
   */
  CertificatoFirma putUltimoCertificatoUsato(String idCertificato);
}

File: cosmowcl/src/app/shared/components/certificati-firma/certificati/certificati.service.ts
/*
 * Copyright CSI-Piemonte - 2023
 * SPDX-License-Identifier: GPL-3.0-or-later
 */
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { CertificatoFirma } from 'src/app/shared/models/api/cosmoauthorization/certificatoFirma';
import { ApiUrls } from 'src/app/shared/utilities/apiurls';

@Injectable()
export class CertificatiService {

  constructor(
    private http: HttpClient
  ) {
    // NOP
  }

  /** GET: Restituisce un certificato. */
  getCertificato(id: number): Observable<CertificatoFirma> {
    return this.http.get<CertificatoFirma>(ApiUrls.CERTIFICATI_ID.replace('{id}', id.toString()));

  }

  /** GET: Restituisce una lista di certificati. */
  getCertificati(): Observable<CertificatoFirma[]> {
    return this.http.get<CertificatoFirma[]>(ApiUrls.CERTIFICATI);
  }

  /** DELETE: Elimina un certificato. */
  deleteCertificato(id: number): Observable<CertificatoFirma> {
    return this.http.delete<CertificatoFirma>(ApiUrls.CERTIFICATI_ID.replace('{id}', id.toString()));
  }

  /**
   * POST: Salva un certificato.
   *
   * @param body CertificatoFirma da salvare.
   * @returns Observable<CertificatoFirma> il certificato salvato.
   */

  salvaCertificato(body: CertificatoFirma): Observable<CertificatoFirma> {
    return this.http.post<CertificatoFirma>(ApiUrls.CERTIFICATI, body);
  }

  /**
   * PUT: Aggiorna un certificato.
   *
   * @param body Certificato da aggiornare.
   * @returns Observable<CertificatoFirma> il certificato aggiornato.
   */

  aggiornaCertificato(id: number, body: CertificatoFirma): Observable<CertificatoFirma> {
    return this.http.put<CertificatoFirma>(ApiUrls.CERTIFICATI_ID.replace('{id}', id.toString()), body);
  }

  salvaCertificatoUltimoUtilizzato(id: number): Observable<CertificatoFirma> {
    return this.http.put<CertificatoFirma>(ApiUrls.CERTIFICATI_ULTIMO_USATO.replace('{id}', id.toString()), {});
  }
}
